src/colmap/estimators/covariance.cc

```diff
@@ -0,0 +1,453 @@
+      for (int j = 0; j < num_params_point; ++j) {
+        H_pp_inv.coeffRef(counter_p + i, counter_p + j) = subMatrix_inv(i, j);
+      }
+    }
+    counter_p += num_params_point;
+  }
+  H_pp_inv.makeCompressed();
+  S_matrix_ = H_cc - H_cp * H_pp_inv * H_pc;
+}
+
+bool BundleAdjustmentCovarianceEstimator::HasValidSchurComplement() const {
+  return S_matrix_.size() != 0;
+}
+
+bool BundleAdjustmentCovarianceEstimator::ComputeFull() {
+  if (!HasValidSchurComplement()) {
+    ComputeSchurComplement();
+  }
+  Eigen::MatrixXd S_dense = S_matrix_;
+  LOG(INFO) << StringPrintf(
+      "Inverting Schur complement for all variables except for 3D points (n = "
+      "%d)",
+      num_params_poses_ + num_params_other_variables_);
+  Eigen::FullPivLU<Eigen::MatrixXd> luOfS_dense(S_dense);
+  if (luOfS_dense.rank() < S_dense.rows()) {
+    LOG(INFO) << StringPrintf(
+        "Unable to compute covariance. The Schur complement on all variables "
+        "(except for 3D points) is rank deficient. Number of columns: %d, "
+        "rank: %d.",
+        S_dense.rows(),
+        luOfS_dense.rank());
+    return false;
+  }
+  cov_variables_ = S_dense.inverse();
+  cov_poses_ = cov_variables_.block(0, 0, num_params_poses_, num_params_poses_);
+  return true;
+}
+
+bool BundleAdjustmentCovarianceEstimator::Compute() {
+  if (!HasValidSchurComplement()) {
+    ComputeSchurComplement();
+  }
+
+  // Schur elimination on other variables to get pose covariance
+  Eigen::MatrixXd S_dense = S_matrix_;
+  LOG(INFO) << StringPrintf("Schur elimination on other variables (n = %d)",
+                            num_params_other_variables_);
+  Eigen::MatrixXd S_aa =
+      S_dense.block(0, 0, num_params_poses_, num_params_poses_);
+  Eigen::MatrixXd S_ab = S_dense.block(
+      0, num_params_poses_, num_params_poses_, num_params_other_variables_);
+  Eigen::MatrixXd S_ba = S_ab.transpose();
+  Eigen::MatrixXd S_bb = S_dense.block(num_params_poses_,
+                                       num_params_poses_,
+                                       num_params_other_variables_,
+                                       num_params_other_variables_);
+  S_bb += lambda_ * Eigen::MatrixXd::Identity(S_bb.rows(), S_bb.cols());
+  Eigen::LDLT<Eigen::MatrixXd> ldltOfS_bb(S_bb);
+  Eigen::MatrixXd S_poses = S_aa - S_ab * ldltOfS_bb.solve(S_ba);
+
+  // Step 5: Compute covariance
+  LOG(INFO) << StringPrintf(
+      "Inverting Schur complement for pose parameters (n = %d)",
+      num_params_poses_);
+  Eigen::FullPivLU<Eigen::MatrixXd> luOfS_poses(S_poses);
+  if (luOfS_poses.rank() < S_poses.rows()) {
+    LOG(INFO) << StringPrintf(
+        "Unable to compute covariance. The Schur complement on pose parameters "
+        "is rank deficient. Number of columns: %d, rank: %d. This is likely "
+        "due to the poses being underconstrained with Gauge ambiguity.",
+        S_poses.rows(),
+        luOfS_poses.rank());
+    return false;
+  }
+  cov_poses_ = S_poses.inverse();
```

Comment:
Can't we instead call `luOfS_poses.inverse()`? This would avoid computing the LU factorization twice.

Reply:
Makes great sense. This is updated. Thanks